### PR TITLE
singlevm : Add networking automation (macvlan and DHCP Server)

### LINKED ISSUE #446: Single VM: Single VM with dnsmasq does not work on Ubuntu
The new single VM setup that uses the bridge with macvlan does not work on ubuntu. The CNCI does not seem to get a IP lease.
Also the dnsmasq command as well as teardown command need sudo to ensure they work.

### PR DESCRIPTION
**Pull request by obedmr**
This change adds the following features to Single VM:

Auto ``macvlan`` interface creation with ip command
DHCP Server setup with ``dnsmasq``

This will help to accomplish #334 and will fix #446

### CODE REVIEW

**Comment by obedmr** [comment]
It's still missing the extra parameter option for Travis running. Something like ``--travis`` in the ``setup.sh`` script in order to skip VMs launching.

**Comment by leoswaldo** [comment]
@obedmr   Coverage decreased (-1.0009%) to 64.457%  that's huge, I'd suggest you to cover that

**Comment by obedmr** [comment]
@leoswaldo I don't thing the issue is on this commit. It's functional testing with bash scripting, any golang code was touched. We need to figure out what's happening with this coverage analysis.

File: testutil/singlevm/setup.sh [hunk 1]
```diff
@@ -63,6 +65,11 @@ sudo killall ciao-scheduler
 sudo killall ciao-controller
 sudo killall ciao-launcher
 sudo killall qemu-system-x86_64
+sudo killall dnsmasq
+echo "Original /etc/hosts is temporarily move to $hosts_file_backup"
```
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr can you document why this needs to be done

File: testutil/singlevm/setup.sh [hunk 2]
```diff
@@ -207,6 +214,27 @@ sudo cp -f clear-"${LATEST}"-cloud.img /var/lib/ciao/images
...
+    sudo ip link set dev eth10 up
+    sudo ip -d link show eth10
+else
+    echo 'ip command is not supported'
+fi
+
+# Set DHCP server with dnsmasq
+sudo mkdir -p /var/lib/misc
+if [ -x "$(command -v ip)" ]; then
+    dnsmasq -C $ciao_scripts/dnsmasq.conf.macvlan0
```
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr this requires sudo

File: testutil/singlevm/setup.sh (hunk 1 above)
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr what if the script hands due to some reason and the user hits Ctrl-C? Then we have potentially messed up the users's system.. may be good to backup the original hosts file at the test directory also

File: testutil/singlevm/setup.sh (hunk 2 above)
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr the dnsmasq file has to be specific for macvlan0. A user can have another dnsmasq instance running. We should run a standalone dnsmasq just for the test and not impact any other dnsmasqs the user may have started

File: testutil/singlevm/setup.sh [hunk 3]
```diff
@@ -327,3 +355,6 @@ fi
 #Also kill the CNCI (as there is no other way to delete it today)
 sudo killall qemu-system-x86_64
 sudo rm -rf /var/lib/ciao/instances
+sudo ip link del eth10
+sudo killall dnsmasq
```
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr can we only kill the one we started.. This should have been done by me earlier. However with your new change, running on the user's system becomes very simple (vs running in a VM) and having this change will help make it more robust.

File: testutil/singlevm/setup.sh [hunk 4]
```diff
@@ -207,6 +214,27 @@ sudo cp -f clear-"${LATEST}"-cloud.img /var/lib/ciao/images
 cd /var/lib/ciao/images
 sudo ln -sf clear-"${LATEST}"-cloud.img df3768da-31f5-4ba6-82f0-127a1a705169
 
+# Set macvlan interface
+if [ -x "$(command -v ip)" ]; then
+    sudo ip link add name eth10 type bridge
```
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr can we delete eth10 before we add it, to handle the issue where the user terminates the script midway with a ctrl C

**Comment by mcastelino** [comment]
@obedmr can we also update the Wiki with new instructions... we can actually simplify the instructions to how run this directly on bare metal. So we can support Bare Metal or Appliance.
https://github.com/01org/ciao/wiki/Single-VM-Machine-Development-Environment

**Comment by mcastelino** [comment]
@obedmr can we add a command line option to only test containers (till @markdryan adds support for a non VT-x qemu launch). Note this can be a separate PR. That will allow us to run this script on travis

**Comment by tpepper** [comment]
Is there an assurance the DHCP server cannot respond to queries from outside the VM?  Ie: it would be bad if our little test appliance running on an arbitrary network was percieved by the upstream network management as a rogue DHCP/DNS source.

**Comment by mcastelino** [comment]
@tpepper the DHCP server is running on a macvlan attached to nothing.. so it should not respond to anything else. Also @obedmr can we make sure we configure it to only respond on the macvlan in the dnsmasq options. I do this in the CNCI today

File: testutil/singlevm/setup.sh (hunk 3 above)
Comment (mcastelino) [inline testutil/singlevm/setup.sh]: @obedmr never mind I see you are already doing that
interface=macvlan0

File: testutil/singlevm/setup.sh (hunk 1 above)
Comment (obedmr) [inline testutil/singlevm/setup.sh]: Yeah, I wanted to backup the hosts files. I'm working on a Ctrl + C trapper. Not a big deal, should be ready fast.

File: testutil/singlevm/setup.sh (hunk 3 above)
Comment (obedmr) [inline testutil/singlevm/setup.sh]: Regarding killing dnsmasq, I changed to killing dnsmasq by PID. I'm getting it from its pidfile

File: testutil/singlevm/setup.sh (hunk 4 above)
Comment (obedmr) [inline testutil/singlevm/setup.sh]: Yeah, removed before creating it.

**Comment by mcastelino** [comment]
LGTM. Not sure why Travis is failing

**Comment by leoswaldo** [comment]
@mcastelino @obedmr  seems to be related with @kaccardi  PR https://github.com/01org/ciao/pull/514

**Comment by obedmr** [comment]
@mcastelino I'll create an issue for not forgetting what's missing. 

- Only-containers option for the setup.sh script
- Ctrl + C trapper function

Also, I'm updating the documentation. Basically, I'd be removing the [Setup Network](https://github.com/01org/ciao/wiki/Single-VM-Machine-Development-Environment#setup-networking) section and add dnsmasq as software dependency.

Is there anything else I should consider?

**Comment by mcastelino** [comment]
@obedmr I wonder if we should have two separate sections in the Wiki.
- Bare Metal Single Machine Test
- VM based Single Machine Test

Specially since bare metal is extremely simple now and something a dev can use with 0 effort.
We need to document bare metal package and hardware and kernel dependencies.

**Comment by mcastelino** [comment]
@obedmr Found one more bug. On the error paths we exit 1. That was not a big issue earlier. However with the new code it will leave dnsmasq around and a corrupted /etc/hosts.
Leaving the dnsmasq around will mean we cannot run the script again.

**Comment by mcastelino** [comment]
LGTM. @tpepper can we merge this in?

**Pull request merged**